proposals/20191030-api.md

```diff
@@ -0,0 +1,118 @@
+# Falco APIs
+
+## Summary
+
+This is a proposal to better structure the Falco API.
+
+It will contain a list of gRPC services that can be added to have a first stable version of the API for v1.0.0.
+
+## Motivation
+
+We want to enable users to use thirdy-party clients to interface with Falco outputs, inputs, rules, and configurations.
+
+Such ability would enable the community to create a whole set of OSS tools, built on top of Falco.
+
+Some examples, already in place, are:
+
+- [falcoctl](https://github.com/falcosecurity/falcoctl)
+- [falco-exporter](https://github.com/falcosecurity/falco-exporter)
+
+## Goals
+
+- To design and implement an API to make Falco more configurable
+- To design and implement an API to make Falco more consumable
+- To design a set of simple contract interfaces
+- To have a secure by default Falco API
+
+## Non-Goals
+
+- Replace the existing inputs: the **inputs** will be only for new source events users may want to add, initially
+
+## Proposal
+
+### Use cases
+
+- Receive Falco events with a well-defined contract over wire --> **outputs**
+- Receive Falco drops with a well-defined contract over wire --> **drops**
+- Receive current Falco version and related meta information (commit hash, built type, etc.) --> **version**
+- Configure Falco via API (CRUD) -> **configs**
+- Inject and/or modify Falco rules via API (CRUD) --> **rules**
```

Comment:
This would be awesome but would still require a route into the kubernetes cluster (or instance) to configure these or a sidecar that is used for configuration.

I am fairly new to the Falco project but has any thought been given to a central management server for Falco agents to report into and fetch new configurations / rules dynamically rather than having to hit an API or reload rules manually? I know this would be a new project to be spun up of the management server but interested if this has been proposed / discussed in the past before

Reply:
Maybe something like [this](https://github.com/falcosecurity/falco/issues/824) can help with what are you suggesting?

Reply:
WRT having any external configuration management, the work starts here with implementing a clean interface around the core of the program. 

I think we would all be very excited to see work like a sidecar or configuration management server come to life, and the Falco community would be happy to sponsor this work. 

If you have ideas for configuration management we could open up a new proposal (like this one) with your ideas, and get a repository and a Falco client set up so work can begin building out something like what you described :smile: